### PR TITLE
List ctrl improvements for different DPIs

### PR DESCRIPTION
**Pull request by gerstrong**
This makes ```wxListView``` to render icons in HIDPI look better.

### CODE REVIEW

File: include/wx/generic/private/listctrl.h
Comment (vadz) [inline include/wx/generic/private/listctrl.h:818]: There shouldn't be any need to store these images here, they're already stored in `m_normal_image_list`, why do we need to duplicate them?

**Review by vadz — CHANGES_REQUESTED** [review]
Sorry, but the changes here seem to be wrong to me. It looks like you're (partially) redoing everything `wxWithImages` already does for you: this class encapsulates the difference between using `SetImageList()` and `SetImages()` for you, you just need to use its `GetImageLogicalSize()` and `GetImageBitmapFor()` methods in the code, what's the problem with doing this?

File: include/wx/generic/private/listctrl.h
Comment (gerstrong) [reply include/wx/generic/private/listctrl.h:818]: @vadz I tried to use ```wxWithImages```. My problem is that you have both normal and small image list objects within one class. If I derive ```wxListMainWindow``` from the class ```wxWithImages``` additionally (like seen for other classes) I would only have one instance I could deal with. What I could do is to create a new class derived from ```wxWithImages``` and create two instances from that (not sure if that would be a good design choice). I also saw ```wxWithImages``` is meant to be not copied (no proper copy-constructors). So I am also a bit unsure about the design of that ```wxWithImages``` on that side. 

Can you guide me a bit on how you think it should be implemented properly seen? I'd be happy to get it coded as intended. At least what I have here works with higher DPI, but only on normal icon size. I tried with small icons but ran into a bunch of crashes, so I left those changes out for now... (but maybe it can be implemented properly as well). I tested this against your samples and it looks good and I also tested it against Money Manager EX and it really gives the application a quality push ;-)

Let me know your thoughts :-)

Comment (vadz) [reply include/wx/generic/private/listctrl.h:818]: We indeed need to do some more refactoring done here, as we must get rid of `m_{small,normal}_image_list` in this class and use `wxListCtrlBase::m_images{Small,Normal}` instead, as we won't be able to deal with high DPI icons correctly as long as `wxImageList` is used directly -- we should only use it via `wxWithImages`. To minimize the number of changes we probably should make the existing `wxImageList*` members `wxWithImages*` instead and make them point to the existing base class objects.

Would this work?

Comment (gerstrong) [reply include/wx/generic/private/listctrl.h]: Okay, I tried your suggestions and so far it seems to work. Hope the tests pass. Let's see...

Comment (gerstrong) [reply include/wx/generic/private/listctrl.h]: Looking good. Let me know your thoughts.

File: include/wx/generic/private/listctrl.h
Comment (vadz) [inline include/wx/generic/private/listctrl.h:698]: I think we should get rid of this function, it seems wrong that it's still used.

File: src/common/listctrlcmn.cpp
Comment (vadz) [inline src/common/listctrlcmn.cpp:295]: Let's please avoid such whitespace-only changes (I can filter it out myself before merging, this is mostly for the future).

File: src/generic/listctrl.cpp
Comment (vadz) [inline src/generic/listctrl.cpp:3333]: We have an overload taking `width` and `height` directly, i.e. this could be just
```suggestion
        m_normal_images->GetImageLogicalSize(this, index, width, height);
```

(and the same thing below, of course).

File: src/generic/listctrl.cpp
Comment (vadz) [inline src/generic/listctrl.cpp:3371]:
```suggestion
        int height;
        images->GetImageLogicalSize(this, 0, width, height);
```

File: src/generic/listctrl.cpp
Comment (vadz) [inline src/generic/listctrl.cpp:3421]: We could minimize the code churn here as well:
```suggestion
            int ix = 0, iy = 0;
            m_small_images->GetImageLogicalSize(this, image, ix, iy);
```

File: src/generic/listctrl.cpp
Comment (vadz) [inline src/generic/listctrl.cpp:4644]: And here:
```suggestion
            int imageWidth, imageHeight;
            m_small_images->GetImageLogicalSize(this, image, imageWidth, imageHeight);
```

**Review by vadz — COMMENTED** [review]
Thanks a lot, this looks much better!

I think we need to get rid of `GetSmallImageList()`, but otherwise I'd be glad to merge it -- although I'd prefer to apply the suggestions below to minimize the unnecessary changes to the code.

TIA!

Note to self: we'd need to avoid using `auto` if we want to backport this 
to 3.2.

**Comment by vadz** [comment]
Looks good to me, thanks! I'll merge this soon.

File: src/generic/listctrl.cpp
Comment (vadz) [inline src/generic/listctrl.cpp:1185]: Just for the record: I'm removing this, it's wrong to use it as a `bool useMask`, and adding `true` here and to all the other places where `DrawBitmap()` is called instead.

File: include/wx/generic/private/listctrl.h
Comment (vadz) [inline include/wx/generic/private/listctrl.h:698]: Also removing this.

**Review by vadz — COMMENTED** [review]
I've finally created #22916 subsuming this PR, so closing it and will merge the other one when the CI builds pass.

**Pull request closed without merging**
